pkgs/applications/virtualization/charliecloud/default.nix

```
@@ -1,35 +1,30 @@
-{ stdenv, fetchFromGitHub, python }:
+{ stdenv, fetchFromGitHub, python3, python3Packages, docker, autoreconfHook, coreutils }:
 
 stdenv.mkDerivation rec {
 
-  version = "0.12";
+  version = "0.18";
   pname = "charliecloud";
 
   src = fetchFromGitHub {
     owner = "hpc";
     repo = "charliecloud";
     rev = "v${version}";
-    sha256 = "177rcf1klcxsp6x9cw75cmz3y2izgd1hvi1rb9vc6iz9qx1nmk3v";
+    sha256 = "0x2kvp95ld0yii93z9i0k9sknfx7jkgy4rkw9l369fl7f73ghsiq";
   };
 
-  buildInputs = [ python ];
+  nativeBuildInputs = [ autoreconfHook ];
+  propagatedBuildInputs = [ python3 docker ] ++ (with python3Packages; [ lark-parser requests ]);
```

This should use `python3.withPackages`. E.g. (untested):

```suggestion
  propagatedBuildInputs = [
    docker
    (python3.withPackages (ps: [ ps.lark-parser ps.requests ]))
  ];
```

Also, if I understand correctly, Python and Docker are runtime dependencies of charliecloud. If so, these should not be in `propagatedBuildInputs` since these dependencies are only available in build environments. You have roughly three options (more or less in order of preference):

1. Put them in `buildInputs`. The configure script picks them up and the Python/Docker binary paths get embedded in the charliecloud binary.
2. Put them in `buildInputs`. They are not picked up, so you have to patch the code where Python/Docker are invoked to use their full paths.
3. Wrap the binary, adding Python/Docker to `PATH`.

---

Yes, you are right...
```
[bzizou@bart:~/git/nixpkgs]$ ./result/bin/ch-grow --list
error: missing dependency: Python module "lark-parser"
error: missing dependency: Python module "requests"
```
Going to fix this... thank you!

---

So, 
1. NOK
2. Don't know how to patch the code for the python packages to be invlolved with their full path...
3. Is there a Nix function to wrap some specific python binaries? or do I have to use `buildPythonApplication` (which requires a `setup.py` file which is not provided here)

---

>     1. NOK

The configure script seems to support `--with-python`. Did you try:

```nix
  configureFlags = let
    pythonEnv = python3.withPackages (ps: [ ps.lark-parser ps.requests ])
  in [
    "--with-python=${pythonEnv}/bin/python3"
  ];
```

You could check if there is a similar flag for Docker.

---

> You could check if there is a similar flag for Docker.

There's not currently, and in fact there's another layer to it: Typically we invoke it as `sudo docker`, so you don't get the user's `PATH`, you get the `PATH` configured for sudo.

This code is lines 147–156 of `lib/base.sh`, and I think that's where you would patch it.

If this flag would be useful to you, please file an upstream feature request and we can look into it. There is interest in pre-processing the shell scripts for other reasons as well (having to do with library paths), but we haven't done it yet.

---

> This code is lines 147–156 of `lib/base.sh`, and I think that's where you would patch it.

Thanks for the pointer. That would need quite some patching, since there a lot of programs are executed that are expected to be in `PATH`, besides `docker` I saw `tar`, `gnused`, `gzip`, and `coreutils` in that file. In pure Nix environments, such packages cannot be expected to be in the environment.

Currently it's probably the easiest to wrap all programs and disable the `sudo` functionality (and expect users to add themselves to the `docker` group).

I will leave it up to @bzizou to file any issues, since I do not know charliecloud.

---

> In pure Nix environments, such packages cannot be expected to be in the environment.

OK, yes, @bzizou let's please have an upstream discussion about how to make your packaging work easier.

---

Well, the `--with-python` with `pythonEnv` seems to work, as ch-grow does not complain anymore about missing libraries.
And regarding docker, I'm a bit surprised about the sudo thing, because I tested an image build and a run, and it works like a charm, maybe just because my login is in the docker group?

Test with Docker:

```
[bzizou@bart:~/tmp/examples]$  ~/git/nixpkgs/result/bin/ch-build -t hello .
building with: docker
Sending build context to Docker daemon  165.9kB
Step 1/3 : FROM debian:stretch
 ---> 2d337f242f07
Step 2/3 : ARG DEBIAN_FRONTEND=noninteractive
 ---> Using cache
 ---> e29f34e1dfdb
Step 3/3 : RUN apt-get update  && apt-get install -y --no-install-recommends apt-utils  && rm -rf /var/lib/apt/lists/*
 ---> Using cache
 ---> 5f6dae1fffc5
Successfully built 5f6dae1fffc5
Successfully tagged hello:latest

[bzizou@bart:~/tmp/examples]$ ~/git/nixpkgs/result/bin/ch-builder2tar hello /var/tmp
builder: docker
exporting
adding environment
compressing
-rw-r--r-- 1 bzizou users 44M Aug 26 13:28 /var/tmp/hello.tar.gz

[bzizou@bart:~/tmp/examples]$ ~/git/nixpkgs/result/bin/ch-tar2dir /var/tmp/hello.tar.gz /var/tmp
replacing existing image /var/tmp/hello
/var/tmp/hello unpacked ok

[bzizou@bart:~/tmp/examples]$ ~/git/nixpkgs/result/bin/ch-run /var/tmp/hello -- echo "I'm in a container"
I'm in a container
```

Test with ch-grow:

```
[bzizou@bart:~/tmp/examples]$  ~/git/nixpkgs/result/bin/ch-grow hello
growing: hello
  3 FROM debian:stretch 
manifest: downloading
layer 1/1: 419e7ae: downloading
layer 1/1: 419e7ae: listing
validating tarball members
resolving whiteouts
flattening image
[...]
```

---

> Well, the `--with-python` with `pythonEnv` seems to work, as ch-grow does not complain anymore about missing libraries.

Probably because you are running in an environment where all the dependencies are available (e.g. a normal user/system profile). It does not work if they are not:

```
$ wrapit --ro-bind /sys /sys --share-net --bind $PWD $PWD `realpath results/charliecloud/bin/ch-builder2tar` hello $PWD/foo
/nix/store/xgn17nidg6lidr2880rlgshjfx1mnbwd-charliecloud-0.18/bin/ch-builder2tar: line 3: dirname: command not found
/nix/store/xgn17nidg6lidr2880rlgshjfx1mnbwd-charliecloud-0.18/bin/ch-builder2tar: line 4: /home/daniel/.cache/nixpkgs-review/pr-95888-1/../lib/charliecloud/base.sh: No such file or directory
```

(`wrapit` calls bubblewrap with some useful binds like `/nix`)

---

@danieldk Many thanks! I learned a lot this days! Bubblewrap is a great tool for doing sandboxing checks!
So, I wrapped the scripts/binaries and created a bash script for testing the package using `bwrap`. The test creates an image with `ch-grow` and runs an echo test inside, as explained in the Charliecloud tutorial.
```
[bzizou@bart:~/git/nixpkgs]$ pkgs/applications/virtualization/charliecloud/test.sh
Temporary image directory: /tmp/tmp.WDmKkgnJe6
Temporary test directory: /tmp/tmp.uxuBVPFXP3
Testing...
growing: hello
  2 FROM debian:stretch
  4 ARG DEBIAN_FRONTEND='noninteractive'
  5 RUN ['/bin/sh', '-c', 'apt-get update  && apt-get install -y --no-install-recommends apt-utils  && rm -rf /var/lib/apt/lists/*']                                                                                                            
Get:1 http://security.debian.org/debian-security stretch/updates InRelease [53.0 kB]
Ign:2 http://deb.debian.org/debian stretch InRelease            
Get:3 http://deb.debian.org/debian stretch-updates InRelease [93.6 kB]
Get:4 http://security.debian.org/debian-security stretch/updates/main amd64 Packages [558 kB]
Get:5 http://deb.debian.org/debian stretch Release [118 kB]        
Get:6 http://deb.debian.org/debian stretch-updates/main amd64 Packages [2596 B]           
Get:7 http://deb.debian.org/debian stretch Release.gpg [2410 B]               
Get:8 http://deb.debian.org/debian stretch/main amd64 Packages [7080 kB]
Fetched 7907 kB in 6s (1281 kB/s)                                                                                      
Reading package lists... Done
W: No sandbox user '_apt' on the system, can not drop privileges
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  libapt-inst2.0
The following NEW packages will be installed:
  apt-utils libapt-inst2.0
0 upgraded, 2 newly installed, 0 to remove and 0 not upgraded.
Need to get 601 kB of archives.
After this operation, 1603 kB of additional disk space will be used.
Get:1 http://deb.debian.org/debian stretch/main amd64 libapt-inst2.0 amd64 1.4.10 [191 kB]
Get:2 http://deb.debian.org/debian stretch/main amd64 apt-utils amd64 1.4.10 [410 kB]
Fetched 601 kB in 0s (1312 kB/s)
debconf: delaying package configuration, since apt-utils is not installed
Selecting previously unselected package libapt-inst2.0:amd64.
(Reading database ... 6503 files and directories currently installed.)
Preparing to unpack .../libapt-inst2.0_1.4.10_amd64.deb ...
Unpacking libapt-inst2.0:amd64 (1.4.10) ...
Selecting previously unselected package apt-utils.
Preparing to unpack .../apt-utils_1.4.10_amd64.deb ...
Unpacking apt-utils (1.4.10) ...
Setting up libapt-inst2.0:amd64 (1.4.10) ...
Setting up apt-utils (1.4.10) ...
Processing triggers for libc-bin (2.24-11+deb9u4) ...
W: No sandbox user '_apt' on the system, can not drop privileges
done; 3 instructions executed
builder: docker
exporting
adding environment
compressing
-rw-r--r-- 1 bzizou users 44M Aug 27 09:00 /tmp/tmp.WDmKkgnJe6/hello.tar.gz
creating new image /tmp/tmp.WDmKkgnJe6/hello
/tmp/tmp.WDmKkgnJe6/hello unpacked ok
I'm in a container. TEST IS OK!
Cleaning
```

---

Awesome! From a quick glance things look to be in a good shape now. I will try to look in more detail at the changes sometime later today.